src/components/DataProvider/index.js

```diff
@@ -18,225 +29,137 @@ export default class DataProvider extends Component {
-    }
-    return false;
-  }
-
-  componentDidUpdate(prevProps, prevState) {
-    if (this.props.loader !== prevProps.loader) {
-      return this.fetchData('NEW_LOADER');
-    }
-    const { baseDomain: domain } = this.props.config;
-    const { baseDomain: oldDomain } = prevProps.config;
-    if (
-      domain &&
-      oldDomain &&
-      (domain[0] !== oldDomain[0] || domain[1] !== oldDomain[1])
-    ) {
-      return this.fetchData('NEW_DOMAIN');
-    }
-    return 1;
-  }
-
   componentWillUnmount() {
     clearInterval(this.fetchInterval);
     this.unmounted = true;
   }
 
+  getSeriesObjects() {
+    const { series, yAccessor, xAccessor } = this.props;
+    const { loaderConfig } = this.state;
+    return series.map(s => ({
+      data: [],
+      ...s,
+      ...loaderConfig[s.id],
+      yAccessor: s.yAccessor || yAccessor,
+      xAccessor: s.xAccessor || xAccessor,
+    }));
+  }
+
   fetchData = async reason => {
-    const series = await this.props.loader(
-      this.props.config.baseDomain,
-      this.state.subDomain,
-      this.props.config,
-      this.state.series,
-      reason
-    );
-    const update = {
-      series,
-    };
+    const { baseDomain, pointPerSeries, defaultLoader } = this.props;
+    const { subDomain } = this.state;
+    const newLoaderConfig = {};
+    const seriesObjects = this.getSeriesObjects();
+    let series = await Promise.map(seriesObjects, async s => {
+      const loader = s.loader || defaultLoader;
+      if (!loader) {
+        throw new Error(`Series ${s.id} does not have a loader.`);
+      }
+      const loaderReturn = await loader(
+        s.id,
+        baseDomain,
+        subDomain,
+        pointPerSeries,
+        seriesObjects,
+        reason
+      );
+      return {
+        ...loaderReturn,
+        id: s.id,
+        reason,
+      };
+    });
+    if (reason === 'MOUNTED') {
+      series = series.map(s => ({
+        ...s,
+        yDomain: calculateDomainFromData(
+          s.data,
+          s.yAccessor || this.props.yAccessor
+        ),
+      }));
+    }
+    const { loaderConfig } = this.state;
+    series.forEach(s => {
+      newLoaderConfig[s.id] = {
+        ...loaderConfig[s.id],
+        ...s,
+      };
+    });
+    const stateUpdates = { loaderConfig: newLoaderConfig };
     if (reason !== 'UPDATE_SUBDOMAIN') {
-      update.contextSeries = series;
-    }
-    if (!this.unmounted) {
-      this.setState(update);
+      stateUpdates.contextSeries = { ...newLoaderConfig };
     }
+    this.setState(stateUpdates);
   };
 
-  subDomainChanged = async subDomain => {
+  subDomainChanged = subDomain => {
     const current = this.state.subDomain;
     if (subDomain[0] === current[0] && subDomain[1] === current[1]) {
       return;
     }
     clearTimeout(this.subDomainChangedTimeout);
-    this.subDomainChangedTimeout = setTimeout(async () => {
-      await this.fetchData('UPDATE_SUBDOMAIN');
-    }, 300);
+    this.subDomainChangedTimeout = setTimeout(
```

Comment:
If the construction with setTimeout was made in order to wait until setState will be finished and the subDomain is set, it is better to put the desired method as a callback, i.e.
```
this.setState({ subDomain }, () => this.fetchData('UPDATE_SUBDOMAIN'))
```

Reply:
Very nice.

Reply:
That's not why though, `setState` is called every time -- then there's a tail function (300 ms after the last `updateSubdomain` call - we ask the loader for data.

Reply:
oh, make sense, than for the code readability I would use lodash throttle